### ISSUE TITLE
How to visualize patch-level classification

### ISSUE THREAD

**Issue by YangJae96** [linyq2117/TagCLIP#5]
Hi. Thank you for great work!.

![Image](https://github.com/user-attachments/assets/b62cae6e-589c-44fe-a8d5-d459bf338849)

How can I visualize the patch-level classification, attention map as above figure??

```
    image_features, attn_weight_list = model.encode_image_tagclip(image, h, w, attn_mask=1)
    image_features = image_features / image_features.norm(dim=-1, keepdim=True)
    

    attn_weight = [aw[:, 1:, 1:] for aw in attn_weight_list]
    
    attn_vote = torch.stack(attn_weight, dim=0).squeeze()
    
    thres0 = attn_vote.reshape(attn_vote.shape[0], -1)
    thres0 = torch.mean(thres0, dim=-1).reshape(attn_vote.shape[0], 1, 1)
    thres0 = thres0.repeat(1, attn_vote.shape[1], attn_vote.shape[2])
    
    if args.dataset == 'cocostuff':
        attn_weight = torch.stack(attn_weight, dim=0)[:-1]
    else:
        attn_weight = torch.stack(attn_weight, dim=0)[8:-1]
    
    attn_cnt = attn_vote > thres0
    attn_cnt = attn_cnt.float()
    attn_cnt = torch.sum(attn_cnt, dim=0)
    attn_cnt = attn_cnt >= 4
    
    attn_weight = torch.mean(attn_weight, dim=0)[0]
    attn_weight = attn_weight * attn_cnt.float()

    logit_scale = model.logit_scale.exp()
    logits = logit_scale * image_features @ text_features.t()#torch.Size([1, 197, 81])
    logits = logits[:, 1:, :]
    logits = logits.softmax(dim=-1)
    logits_coarse = logits.squeeze()
    logits = torch.matmul(attn_weight, logits)
    logits = logits.squeeze()
    logits = mask_attn(logits_coarse, logits, h, w, attn_weight)

    logits_max = torch.max(logits, dim=0)[0]
    logits_max = logits_max[:NUM_CLASSES]
    logits_max = cwr(logits, logits_max, h, w, image, text_features)
    logits_max = logits_max.cpu().numpy()
pred_label_id.append(logits_max)
```

The code is from classify.py.

Should I use the logits   in the code for visualization?

If possibile, providing some code will be really thankful.!

**Comment by linyq2117** [comment]
Thanks for your interest!

Yes, the visualization can be realized by using the logits in the code. Unfortunately, the visualized code is no longer available. We have provided pseudocode as a reference for your own implementation. 

```
def colorful(out, name):
    arr = out.astype(np.uint8)
    im = Image.fromarray(arr)

    palette = []
    for i in range(256):
        palette.extend((i, i, i))

    palette[:3 * 21] = np.array([[0, 0, 0],
                                 [128, 0, 0],
                                 [0, 128, 0],
                                 [128, 128, 0],
                                 [0, 0, 128],
                                 [128, 0, 128],
                                 [0, 128, 128],
                                 [128, 128, 128],
                                 [64, 0, 0],
                                 [192, 0, 0],
                                 [64, 128, 0],
                                 [192, 128, 0],
                                 [64, 0, 128],
                                 [192, 0, 128],
                                 [64, 128, 128],
                                 [192, 128, 128],
                                 [0, 64, 0],
                                 [128, 64, 0],
                                 [0, 192, 0],
                                 [128, 192, 0],
                                 [0, 64, 128]
                                 ], dtype='uint8').flatten()
    im.putpalette(palette)
    im.save(name)

# logits (B, N, C)
logits = logits.transpose(-2, -1).reshape(1, C, H, W)
seg_vis = F.interpolate(logits, size=(ori_height, ori_width), mode="bilinear")
seg_vis = seg_vis.squeeze()
seg_vis = torch.argmax(seg_vis, dim=0) + 1
seg_vis[seg_vis > 20] = 0 # For VOC, ignore bg categories.
colorful(seg_vis, save_path)
```

**Comment by linyq2117** [comment]
Closed for inactivity.
